plugin/completion.py

```diff
@@ -72,274 +111,81 @@ def initialize(self) -> None:
+        if not self.initialized:
+            self.initialize()
 
-    def is_same_completion(self, prefix: str, locations: List[int]) -> bool:
-        if self.response_incomplete:
-            return False
+        if not self.enabled:
+            return None
 
-        if self.last_location < 0:
-            return False
+        completion_list = sublime.CompletionList()
 
-        # completion requests from the same location with the same prefix are cached.
-        current_start = locations[0] - len(prefix)
-        last_start = self.last_location - len(self.last_prefix)
-        return prefix.startswith(self.last_prefix) and current_start == last_start
-
-    def find_completion_item(self, inserted: str) -> Optional[dict]:
-        """
-
-        Returns the completionItem for a given replacement string.
-        Matches exactly or up to first snippet placeholder ($s)
-
-        """
-        # TODO: candidate for extracting and thorough testing.
-        if self.completions:
-            for index, item in enumerate(self.completions):
-                trigger, replacement = item
-
-                snippet_offset = replacement.find('$', 2)
-                if snippet_offset > -1:
-                    if inserted.startswith(replacement[:snippet_offset]):
-                        return self.response_items[index]
-                else:
-                    if replacement == inserted:
-                        return self.response_items[index]
-        return None
-
-    def on_modified(self) -> None:
-
-        # hide completion when backspacing past last completion.
-        if self.view.sel()[0].begin() < self.last_location:
-            self.last_location = -1
-            self.view.run_command("hide_auto_complete")
-
-        # cancel current completion if the previous input is an space
-        prev_char = self.view.substr(self.view.sel()[0].begin() - 1)
-        if self.state == CompletionState.REQUESTING and prev_char.isspace():
-            self.state = CompletionState.CANCELLING
-
-        if self.committing:
-            self.committing = False
-            self.on_completion_inserted()
+        # this is for tests
+        if self.test_completions:
+            self.handle_response(self.test_completions, completion_list, prefix)
         else:
-            if self.view.is_auto_complete_visible():
-                if self.response_incomplete:
-                    # debug('incomplete, triggering new completions')
-                    self.view.run_command("hide_auto_complete")
-                    sublime.set_timeout(self.run_auto_complete, 0)
+            self.do_request(completion_list, prefix, locations)
 
-    def on_completion_inserted(self) -> None:
-        # get text inserted from last completion
-        begin = self.last_location
+        return completion_list
 
-        if begin < 0:
+    def do_request(self, completion_list: sublime.CompletionList, prefix: str, locations: List[int]) -> None:
```

Comment:
So as I mentioned before, let's focus on getting the basics right and not support multiple cursors for now.